build-tools/scripts/jenkinsfile

```diff
@@ -0,0 +1,484 @@
+// This file is based on the Jenkins declarative pipeline syntax (as opposed to the scripted pipeline syntax)
+// https://jenkins.io/doc/book/pipeline/syntax/
+// https://jenkins.io/blog/2017/02/03/declarative-pipeline-ga/
+
+def botLabel = 'xamarin-android-master-pipeline-eval'
+
+def XADir = "xamarin-android"
+
+def publishBuildFilePaths = null
+def publishTestFilePaths = null
+
+def MSBUILD_AUTOPROVISION_ARGS="/p:AutoProvision=True /p:AutoProvisionUsesSudo=True /p:IgnoreMaxMonoVersion=False"
+
+def buildStatusProcessed = false
+def buildStatusUploaded = false
+
+def isPR = false                // Default to CI
+
+def processBuildStatus() {
+    def status = 0
+
+    try {
+        echo "processing build status"
+        sh "make package-build-status CONFIGURATION=${env.BuildFlavor}"
+    } catch (error) {
+        echo "ERROR : NON-FATAL : processBuildStatus: Unexpected error: ${error}"
+        status = 1
+    }
+
+    return status
+}
+
+def publishPackages(filePaths) {
+    def status = 0
+    try {
+        echo "Publishing to Azure storage account: https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/7b4817ae-218f-464a-bab1-a9df2d99e1e5/resourceGroups/Group/providers/Microsoft.ClassicStorage/storageAccounts/xamjenkinsartifact/overview"
+        step([$class: 'WAStoragePublisher', allowAnonymousAccess: true, cleanUpContainer: false, cntPubAccess: true, containerName: "${env.ContainerName}", doNotFailIfArchivingReturnsNothing: false,
+            doNotUploadIndividualFiles: false, doNotWaitForPreviousBuild: true, excludeFilesPath: '',
+            filesPath: "${filePaths}", storageAccName: "${env.StorageAccountName}", storageCredentialId: "${env.StorageCredentialId}",
+            uploadArtifactsOnlyIfSuccessful: false, usploadZips: false, virtualPath: "${env.StorageVirtualPath}", storageType: 'blobstorage'
+        ])
+
+    } catch (error) {
+        echo "ERROR : publishPackages: Unexpected error: ${error}"
+        status = 1
+    }
+
+    return status
+}
+
+def cleanFiles(path) {
+    // path can include wildcard characters
+
+    def status = 0
+
+    try {
+
+        script {
+            sh """
+            # Note: Escape dollar \$ even in bash comments to avoid groovy templating errors; use non-escaped dollar to expand groovy variables
+            for file in `ls ${path}`; do
+                [ -f "\$file" ] || continue
+                echo "Cleaning file '\${file}''"
+                rm -f "\$file"
+            done
+            """
+        }
+
+    } catch (error) {
+        echo "ERROR : cleanFiles: Unexpected error: ${error}"
+        status = 1
+    }
+
+    return status
+}
+
+def cleanDirectories(path) {
```

Comment:
this seems quite complicated, why is a normal `rm -rf` not sufficient?

Reply:
Addressed with commit https://github.com/xamarin/xamarin-android/commit/aa03f1a92195a2860e0db9e89da349e50a941972
The xamarin-android freestyle build includes a git behavior 'clean after checkout' included in the Jenkins build definition.  The pipeline build definition also has this option and so I enabled it. So I'm able to do away with the entire cleaning business.